### DISCUSSION TITLE
Improving Registration Speed

### DISCUSSION THREAD

**Discussion by cruff3** [uncbiag/uniGradICON#40] [Q&A]
Hello @HastingsGreer and others,

When I use uniGradICON for registering two images, the default speed (on my end) is around 1-1:30 minutes. While this is relatively quick, I need to perform many registrations in a batch (i.e. 50+ registrations per patient), which results in ~1 hour long total computational time per patient. Do you (or anyone) have any tips on how to improve computation time?

Thanks ahead of time!

**Comment by lintian-a** [comment]
> Hello @HastingsGreer and others,
> 
> 
> 
> When I use uniGradICON for registering two images, the default speed (on my end) is around 1-1:30 minutes. While this is relatively quick, I need to perform many registrations in a batch (i.e. 50+ registrations per patient), which results in ~1 hour long total computational time per patient. Do you (or anyone) have any tips on how to improve computation time?
> 
> 
> 
> Thanks ahead of time!

You can try to turn the IO off or use less iterations for the IO stage, say about 10 iterations. Without IO, the runtime probably will be within 1 second excluding the loading time of the model. 

The speed could be faster by running the inference(w/o IO) with batch. But this requires development and modification of the command line interface to accept a list of input files. I will add this in my todo list but cannot promise when the new interface will be released.

**Reply by cruff3** [reply]
Hi @lintian-a ,

Thanks for your reply! I am currently using zero IO iterations, and when using cuda, the registration time is ~30 seconds per registration. While this is an improvement to before, it still takes quite a bit of time to perform all the registrations. It seems that the limiting step is the "loss = model(A_resized, B_resized)" line, where A and B are passed into the GradientICONSparse function. Even though the loss is not used with zero IO iterations, it seems that A_resized and B_resized must still be passed to GradientICONSparse to calculate phiAB (commenting out the "loss..." line leads to an error where phiAB is undefined). Do you know if there are any ways to make this faster/change the class to avoid propagating loss (or whatever the limiting step is)? For my purposes, registration speed is more important than fine-tuning the model, as the registrations are already very good.

Thanks again!

**Reply by lintian-a** [reply]
loss = model(A_resized, B_resized) asks the model to compute the transformation given the paired images. Thus, it is necessary. One workaround would be to comment out the computation of loss inside the GradientICONSparse. You probably can do it by commenting out the rest after [line 47](https://github.com/uncbiag/uniGradICON/blob/1ca39a817e26c92c30266dadf1b5b971f88f253f/src/unigradicon/__init__.py#L47) and return None instead of the ICONLoss.

Loading the weight would also take some time. Thus, customizing the inference code to load the model weight once and run the inference on a list of paired images would save you a lot of time, especially if you want to run it on a large dataset.
